### ISSUE TITLE
Stripe Issue No such customer: cus_xxxxxxxxxxx

### ISSUE THREAD

**Issue by philipboomy** [craftcms/commerce-stripe#62]
### Description
When clicking "pay" I get this error: No such customer: cus_xxxxxxxxxxx - It only happens if I have started checking out and then log into the admin area and then go back to checking out. I can see it merges my cart or something as I suddenly have more products in the cart. I am using the "Stripe Payment Intents" setting but testing with normal 4242 4242 4242 4242 card which does not "force" a 3D check

### Steps to reproduce

1. Add to cart
2. Go to checkout using example templates 
3. Check out as guest
4. Sign in to the admin area - Perhaps via a new tab in the browser
5. Notice that the cart items changes (if your login user has items in the cart)
6. Fill out Stripe credit card and click pay.
7. Error appears

### Additional info

- Craft version: 3.1.30 (but happened on previous versions too)
- PHP version: 7.2.10
- Database driver & version: PostgreSQL 10.5
- Plugins & versions: Craft Commerce 2.1.7, Stripe for Craft Commerce 2.0.1.1

**Comment by fantasticmachine** [comment]
Are you using Stripe Charge or Stripe Payment Intents?
I get that issue when using Payment Intents but not with Stripe Charge.

**Comment by philipboomy** [comment]
Payment Intents due to SCA

**Comment by fantasticmachine** [comment]
Yeah I've been looking at this on and off for a couple of weeks and I'm getting more convinced there's an issue with the plugin. Either that or something important isn't documented.

Are you using the built in card payment fields or rolling your own?

**Comment by philipboomy** [comment]
I have tried both, built in and my own version, and I get the issue on both.

**Comment by andris-sevcenko** [comment]
Sounds like you're using a Craft user that had a Stripe customer associated to it which then got deleted on Stripe's end, but not Commerce's.

**Comment by fantasticmachine** [comment]
So is the stripe customer ID stored against the Commerce customer somewhere? Is it necessary for me to delete the customer account entirely from Commerce?
I've tried this with completely randomly generated users and still get the following error in Commerce.

> Invalid payment or order. Please review.

The payment does not appear in Stripe. However the Stripe Event log shows a POST /v1/sources event which appears to have been successful.

I'm trying to trawl the Craft log at the moment to see if there is more information on the error. But when I encountered this last week I was getting the same error as @philipboomy mentions.

**Comment by andris-sevcenko** [comment]
> So is the stripe customer ID stored against the Commerce customer somewhere? Is it necessary for me to delete the customer account entirely from Commerce?

Can you check the `craft_stripe_customers` table to find the matching customer reference?

> The payment does not appear in Stripe. However, the Stripe Event log shows a POST /v1/sources event which appears to have been successful.

The failed attempt would probably show up as `POST /v1/payment_intents`.

If you find that there's an entry in your `craft_stripe_customers` table that has no match on Stripe's end, you can just delete that row in the local table. I'll see if I can come up with a more elegant solution.

**Comment by fantasticmachine** [comment]
The `craft_stripe_customers` table is completely empty

The only events that show up are for `POST /v1/sources` so that makes me think something is wrong with my front end implementation

**Comment by andris-sevcenko** [comment]
Huh. Is this in a site where it's possible to get FTP/CP access to? If so, can you send that over to support@craftcms.com and reference this issue?

**Comment by fantasticmachine** [comment]
I may be able to do that, sure. Will just check the FTP situation
Just updated my template to use the payment intents JS from the plugin.
Now I get a `POST /v1/payment_methods` entry in the Stripe log but still the same error reported by Commerce `Invalid payment or order. Please review.`

**Comment by philipboomy** [comment]
I got a customer in craft_stripe_customers and it is userId 1 and it got that "cus_FA58z20tu7liHA" in the response column. So I think you are right. Thats the issue. Does this help you ? Do you want access to the site?

**Comment by andris-sevcenko** [comment]
@philipboomy see if nuking that row resolves it for you.

**Comment by andris-sevcenko** [comment]
@fantasticmachine you'd get that error if either the order or the payment form failed to validate. What payment form HTML are you using? If you just use `gateway.getPaymentForm()` does that fix it for you?

**Comment by philipboomy** [comment]
@andris-sevcenko nuked it and it worked

**Comment by fantasticmachine** [comment]
@andris-sevcenko Yep switching to use `{{ cart.gateway.getPaymentFormHtml({})|raw }}` works
So I need to look at what's going on in my payment form I guess.

**Comment by andris-sevcenko** [comment]
@fantasticmachine (it's pretty simple, really: https://github.com/craftcms/commerce-stripe/tree/develop/src/templates/paymentForms)

**Comment by fantasticmachine** [comment]
@andris-sevcenko I think the cause of the problem was that I had `data-enforce3dsecure="{{ cart.gateway.enforce3dSecure ?? '' }}"` on my fields container. When I remove that, it seems to work.

I'd pretty much just copied the fields and JS from the plugin anyway but without the CSS classes and formatted to use our standardised checkout process.

Going to do some testing but it's looked good so far.

Thanks for the help

**Comment by santanup789** [comment]
I'm testing the payments from 2accounts. From 1account all the steps are cool. But the original one which will be use for commercial purpose that's making noise. No such customer: cus_XXXXXXXXXXXX

**Comment by j-greig** [comment]
I've just encountered the "No such customer" error and can see a user reference in the stripe_customers table for it... could someone confirm what is the correct way to avoid this happening?

(My first payment using a 3D Secure test card number went through successfully (using a registered and logged-inuser in Commerce) but the second attempt errored... and it seems to work fine if I'm an unregistered customer.

**Comment by hotcakedev628** [comment]
I am facing this issue, too.
In my back-end process, first `retrieve` a customer already stored in the database, if it doesn't return customer id, then I am going to create a new customer.
What if I just `create` a customer without `retrieve`?

**Comment by fantasticmachine** [comment]
You may be getting the no such customer error if you are using a database from a site that is connected to a different Stripe account.

What this could mean is that the site believes you have a customer record on Stripe. But that record doesn't exist in the Stripe account to which you are posting.

For instance, you have a staging site that is connected to a client's Stripe account and posting to it in dev mode.
You grab a copy of that database to use locally for development. But your local development site is using your own Stripe account in dev mode. Therefore, when your Commerce installation talks to Stripe and passes the customer ID, Stripe has no record that matches and returns an error.

**Comment by jonathanmelville** [comment]
@fantasticmachine I realize I'm really late to this conversation, but I think what you described above is the source of the issue I'm having. I pulled a copy of the live DB and pulled it down to local dev so I could test a Commerce upgrade. I'm getting the "No such customer" error.

So if this is caused by pulling a live DB and running it locally in development, what is the proper way we are supposed to test Commerce upgrades?

**Comment by fantasticmachine** [comment]
Hey @jonathanmelville. When I have to do this I generally delete the stripe_customers table. Unfortunately this means that any tests you do will create new stripe customers in your Stripe Test environment but for me that generally doesn't matter.
It would be awesome if it were possible to clone data from live to test environment in Stripe itself but I don't think that's possible and it could well lead to privacy issues.

**Comment by lukeholder** [comment]
If Stripe had a way to clone their production data to their test mode then this wouldn't be as hard to deal with, since you could use test stripe keys in local dev with a copy of your production database to test an upgrade etc.

We introduced the reset data command to make it a little easier to reset commerce specific data though: https://craftcms.com/docs/commerce/3.x/console-commands.html#reset-data

**Comment by hiasl** [comment]
@lukeholder couldn't you add a hash of the stripe keys to the stripe_customers table to identify the customer based on the stripe key that was used?
